### ISSUE TITLE
Improve the Configuration GUI

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#853]
This is a pretty broad topic that's come up many times before.

In general we don't have any more real-estate to squeeze more config options in the existing Configuration GUI (aka property sheet pages), so this results in overloaded UI (eg. #852) or configuration only being available from the command line (eg. #712) and so affects discoverability too.

Also a simple interface for selecting cards in the 8 slots would fall under this topic too.

**Comment by Keatah** [comment]
I feel that AppleWin will need to expand its GUI at some point - if it is to keep growing. I don't mind the command-line and all, but there are some, maybe even many, options that need their own buttons and tabs. The RamWorks memory bank options come to mind at the moment.

To the right of the ADVANCED tab there looks like there's room for a new ADVANCED-2 tab. Or perhaps 2 or 3 "Aux1", "Aux2", and Aux3 tabs. Perhaps the "Disk" tab could be made a bit smaller to widen the free space. There looks to be 2 or 3 wasted spaces after the "k" in "Disk".

I even see some real estate  on the "Sound" tab. Don't know what would go there. But it's there.

Emulators are technical programs which mimic vintage computers.They have to cover a lot of ground. Especially when they cover expandable systems like the Apple II and Atari 400/800. And this means options. Not only that, but everyone uses emulators in subtly different ways. The more customization offered by the GUI the better.

A busy menu isn't a big deal. We like it that there are buttons to play with & tweak. Once you've learned the emulator and need to change something, it becomes second nature. All options are tuned out except the one you're looking for. And us users appreciate an accommodating program, Altirra is a good example. Every release (even the interim beta-tests) get new options.

And with today's high-resolution displays and uber-widescreen monitors, there's really no need to cram everything into one menu page smaller than the "Disk Properties" in Windows.

**Comment by Keatah** [comment]
Make no mistake, we all love AppleWin over here and appreciate the time and effort put into it. Not to mention it's heritage going back some 20+ years already. This and a few other select emulators have support and continuity rivaling and exceeding many commercial products.

It's used weekly for all kinds of vintage-ee things. Game-playing, curating/archiving, relaxing with BASIC programming. All sorts of good stuff.

**Comment by tomcw** [comment]
Thanks for sharing your UI thoughts and also your positive feedback on AppleWin :+1: 

When we extend & update the UI, I wouldn't want to stick with this out-dated tabbed-property sheet UI, which doesn't scale well for >5 sheets (as you get 2 rows of tabs, with the tabs switching between rows... IMO very confusing!).

Instead, _perhaps_ a tree-control on the left, which select a property sheet on the right, eg. like Visual Studio's "Options":

![VS2019-Options](https://user-images.githubusercontent.com/6696896/107887076-7a24a280-6efb-11eb-8dfe-a0662a57fa5c.png)

Anyway just an idea at this stage.

**Comment by Keatah** [comment]
I think the VisualStudio style of options looks great and would function well.

**Comment by tomcw** [comment]
From the 2021 Emulation Evaluation in Juiced.GS [here](https://juiced.gs/emulators/#applewin), focussing on AppleWin:

>Many more hardware configuration options are available by launching AppleWin from a command prompt or batch file, or by appending the appropriate command line switches to the target of a Windows shortcut. While not all these are of everyday interest, it's unfortunate that this method is required for potentially desirable configurations, such as more than 128K of memory, or a default peripheral card removed from its slot. It would benefit AppleWin if some of the command line launch options were added to its configuration window. Still, it's nice to have them exist at all.

(There is a 2022 [update](https://juiced.gs/store/volume-27-2022/), but it's behind a paywall)

**Comment by Michaelangel007** [comment]
> Instead, _perhaps_ a tree-control on the left, which select a property sheet on the right, eg. like Visual Studio's "Options":

Yes. Unfortunately there really aren't any good ways to manage lots of options; the UI will always end up cluttered with the user scrolling & searching for options. 3D modeling programs are probably the "best" stress test demonstrating this.

A vertical tabbed tree style is "good enough" and solves the immediate needs of providing _someplace_ for users to _start_ customization instead of having to use a clunky [1] command line.

As long as we have a search text field and good categories this is probably the best way to go.

[1] Clunky for non-power users.

**Comment by skiphipdata** [comment]
I cast my vote for the tree-control style GUI. The WinUAE Amiga emulator, which has astronomically more configuration options than any Apple II emulator ever possibly could, uses this type of GUI to great effect, with a smart combination of check boxes, drop-down lists, radio buttons, sliders, etc., with each page consisting of several logical groups of related outlined controls.

I also suggest a panel for slot/card configuration that contains a series of drop-down menus, one for each slot. For each slot, there would be a drop-down containing a list of the cards that can go in that slot, as well as an "Empty" option. I think there is too much inconsistency and clutter with the slot options spread out across different tabs. The set of drop-downs could look like the following:

Slot 0:		Empty
			Language Card

Slot 1:		Empty
			Parallel printer card

Slot 2:		Empty
			Super Serial Card

Slot 3:		Empty
			VidHD
			4Play
			SNES MAX

Slot 4:		Empty
			Mouse interface
			CP/M SoftCard
			4Play
			SNES MAX
			Phasor
			Mockingboard

Slot 5:		Empty
			CP/M SoftCard
			4Play
			SNES MAX
			SAM/DAC
			Mockingboard
			Disk ][

Slot 6:		Empty
			Disk ][

Slot 7:		Empty
			Hard disk controller

Of course, some logic would need to be there to prevent inserting the same card into multiple slots where it doesn't make sense (such as CP/M SoftCard in both slots 4 and 5.) And possibly any configurable options pertaining to a particular card could appear upon selection in a panel to the right of the card selector.

The agat Apple II emulator for Windows implements a slot configuration system along these lines already, if you want to see an example. And the Virtual ][ emulator for macOS also handles slot configuration quite elegantly and attractively this way.

This would have the added benefit of not requiring the command line to remove certain cards (Disk ][, SSC, etc.) from the configuration.

**Comment by tomcw** [comment]
WIP... I've re-purposed the "Sound" configuration property sheet page, and I'm experimenting with a new "Slots" page:
(I've shelved the tree-control idea (above) for now, as the "Slots" page is quicker/easier to implement.)

<img width="340" height="491" alt="Image" src="https://github.com/user-attachments/assets/ff5dfaba-6894-48df-9ded-700dd2afe519" />

I've removed most of the restrictions, so now cards can be in any slot - with just a few restrictions (*)

(*) Restrictions:
- Apple II/II+ Language Card only in slot 0
- Single instance of: Parallel, Serial, Mouse, 4Play, SNESMAX, Uthernet, Uthernet2, VidHD, Z80
  - (or multiple instance of: Disk][ I/F, HDD I/F, Saturn, Mockingboard, Phasor, SAM)
- Can't have both Uthernet and Uthernet2 inserted

The `...` after each slot, is for options unique to that card, eg:

Disk][
<img width="319" height="180" alt="Image" src="https://github.com/user-attachments/assets/871ee3ef-ffc4-4bb1-9d97-187f89f59d06" />

Uthernet(2)
<img width="409" height="284" alt="Image" src="https://github.com/user-attachments/assets/710df2cf-a0fb-448c-a61a-37c49d27a801" />

Hopefully I can share a test build soon.

**Comment by sicklittlemonkey** [comment]
Nice work Tom. That would free up room on some other tabs. (Maybe even replace the Disk one.)

**Comment by tomcw** [comment]
The new Config UI is now in the main line.
If you find any new issues or regressions in AppleWin's new Config UI, then please report them in #1470.

**Comment by tomcw** [comment]
I've made a pre-release (beta) [here](https://github.com/AppleWin/AppleWin/releases/tag/v1.31.0.2), if anyone wants to give it a try.

**Comment by skiphipdata** [comment]
I tried out your beta and am quite happy with it. It's pretty much exactly what I was wanting. It puts all of the slot-based peripheral configuration in one place, instead of it being scattered throughout the config dialog. It also allows for much easier and logical slot configuration than using the command like arguments, which were starting to get unwieldy.

For the sake of completeness, do you think you could add the few slot-related items that are still available by command line but not yet in the GUI? These are:

- Add a Saturn64 option for Slot 0 for Apple][/][+ (or a configuration breakout for Saturn to select desired memory size)
- SSI263 and SC01 speech chip options for Mockingboard/Phasor sockets 0 and 1
- 13-sector option for Disk][ cards
- Checkbox to enable/disable No-Slot-Clock

I think the items above would make the GUI complete WRT slot card configurations. Thanks so much for implementing this! Looking great so far!

**Comment by tomcw** [comment]
Only "quite" happy? ;-)
See: https://languagesandmore.com/quite-vs-quite-uk-vs-us-english/

>For the sake of completeness, do you think you could add the few slot-related items that are still available by command line but not yet in the GUI?

Yes, all good suggestions (and there will be others too) - I'd probably do the config breakout for Saturn memory size.

**Comment by skiphipdata** [comment]
I mean "quite" in the American sense. I'm very happy with the new GUI :)
Having studied linguistics a fair bit, I appreciate that you pointed out the regional semantics of the word "quite."

**Comment by tomcw** [comment]
I've made a 2nd pre-release (beta) [here](https://github.com/AppleWin/AppleWin/releases/tag/v1.31.0.3) - please give it a try.
(@skiphipdata - it includes your above suggestions)

**Comment by skiphipdata** [comment]
Looks great to me!

**Comment by tomcw** [comment]
Closing issue, as this issue is *finally* fixed and available in this new [1.32.0.0 build](https://github.com/AppleWin/AppleWin/releases/tag/v1.32.0.0).
